app/controllers/omniauth_callbacks_controller.rb

```diff
@@ -1,14 +1,21 @@
 class OmniauthCallbacksController < Devise::OmniauthCallbacksController
   include Devise::Controllers::Rememberable
 
-  # Called upon successful redirect from Twitter
-  def twitter
-    callback_for(:twitter)
-  end
-
-  # Called upon successful redirect from GitHub
-  def github
-    callback_for(:github)
+  # Rails actionpack only allows POST requests that come with an ORIGIN header
+  # that matches `request.base_url`, it raises CSRF exception otherwise.
+  # There is no way to allow specific ORIGIN values in order to securely bypass
+  # trusted origins (i.e. Apple OAuth) so `protect_from_forgery` is skipped
+  # ONLY when it's safe to do so (i.e. ORIGIN == 'https://appleid.apple.com').
+  # The hardcoded CSRF check can be found in the method `valid_request_origin?`:
+  # https://github.com/rails/rails/blob/master/actionpack/lib/action_controller/metal/request_forgery_protection.rb#L453-L463
+  protect_from_forgery unless: -> { safe_apple_callback_request? }
```

Comment:
Okay so I gave up on `rack-cors`, I now understand how this works 😅 Here's the gist:

1. `rack-cors` sets the CORS headers and these headers are used by the browser to allow/disallow requests. Our problem was not coming from here
2. The CSRF check is done in Rail's `actionpack` [here](https://github.com/rails/rails/blob/master/actionpack/lib/action_controller/metal/request_forgery_protection.rb#L453-L463), notice how __there is no way to "safely override" this__. So I added this check to bypass `protect_from_forgery` only when it's safe to do so.

Basically, we only want to "open our door" when Apple (trusted origin) is knocking, instead of bypassing this altogether ([see this comment](https://github.com/forem/forem/pull/7952#discussion_r466851860) for the backstory of the analogy)

Also, check this out as it's now deployed and should be working in https://rpi-forem.ngrok.io/

Reply:
This is a bit crazy, it worked for me yesterday. Now that I'm trying it again today I got the same error as you (CSRF), but after I clear my caches and service worker (basically delete all Application data) it works again :(

I've seen this happen a few times now and I think something in the service worker may be interfering with the flow. Can you maybe try it again after clearing the caches?

I think Apple's redirect may be cached by the service worker, I worry we'll see the same result in a real deployment but it may be good to test it out. I'll try to see if there's something we can avoid to cache over in the service worker JS

Reply:
Booted up Safari from scratch.
Tried the first time with the app already connected to the rpi-forem and it failed. Went to appleid.apple.com, disconnected the app, didn't work. 

Tried again from Chrome, it worked!